release_dashboard/lib/widgets/create_release_substeps.dart

```diff
@@ -112,30 +140,67 @@ class CreateReleaseSubstepsState extends State<CreateReleaseSubsteps> {
 }
 
 typedef SetReleaseData = void Function(String name, String data);
+typedef ChangeIsEachInputValid = void Function(int index, bool isValid);
 
 /// Captures the input values and updates the corresponding field in [releaseData].
 class InputAsSubstep extends StatelessWidget {
-  const InputAsSubstep({
+  InputAsSubstep({
     Key? key,
     required this.index,
     required this.setReleaseData,
     this.hintText,
+    required this.changeIsInputValid,
   }) : super(key: key);
 
   final int index;
   final SetReleaseData setReleaseData;
   final String? hintText;
+  final ChangeIsEachInputValid changeIsInputValid;
 
   @override
   Widget build(BuildContext context) {
+    late RegExp formRegexValidator;
+    // assigns corresponding regex to each input
+    switch (index) {
+      case 0:
+        formRegexValidator = candidateBranchRegex;
+        break;
+      case 2:
+      case 3:
+        formRegexValidator = githubRemotePattern;
+        break;
+      case 4:
+      case 5:
+        formRegexValidator = multiGitHashRegex;
+        break;
+      case 6:
+        formRegexValidator = gitHashRegex;
+        break;
+      default:
+        break;
+    }
+
     return TextFormField(
       key: Key(CreateReleaseSubsteps.substepTitles[index]),
+      autovalidateMode: AutovalidateMode.always,
       decoration: InputDecoration(
         labelText: CreateReleaseSubsteps.substepTitles[index],
         hintText: hintText,
       ),
       onChanged: (String data) {
         setReleaseData(CreateReleaseSubsteps.substepTitles[index], data);
+        if (!formRegexValidator.hasMatch(data)) {
+          changeIsInputValid(index, false);
+        } else {
+          changeIsInputValid(index, true);
+        }
+      },
+      validator: (String? value) {
+        if (!formRegexValidator.hasMatch(value ?? '')) {
+          return '${CreateReleaseSubsteps.substepTitles[index].replaceAll(' (if necessary)', '')} not in a valid format!';
```

Comment:
Isn't this value already accessible in scope as `value`?

Reply:
Also, once you refactor this into multiple classes for the different types of input, I think it would be nice to hardcode in an example of the valid input. Something like, `"This must be a valid candidate branch string, e.g. 'flutter-1.2-candidate.3'"`

Reply:
> Response from Alex: `CreateReleaseSubsteps.substepTitles[index]` is the title of the input element, and `value` is the input the user provides.

Ahh, I see, you're right.